docs/plans/worker-flow-run-cancellation.md

```diff
@@ -0,0 +1,457 @@
+
+1. **Startup Poll**: On worker startup, query the API for any flow runs in `CANCELLING` state belonging to this work pool. This catches orphaned flow runs from before the worker started (e.g., after a worker restart).
+
+2. **WebSocket Events**: Subscribe to `prefect.flow-run.Cancelling` events for real-time detection of new cancellations.
+
+No continuous polling is needed—the startup poll handles the restart case, and WebSocket events handle the steady-state case.
+
+### Force Cancellation Flow
+
+When the Worker detects a cancelling flow run:
+
+1. Parse `infrastructure_pid` to verify it's infrastructure this worker can manage
+2. Calculate time remaining until grace period expires based on `state.timestamp`
+3. Schedule an async task to check again after the remaining time
+4. When the task fires, re-fetch the flow run state
+5. If still `CANCELLING`, call `kill_infrastructure()` and mark as `CANCELLED`
+6. If already `CANCELLED` or other terminal state, do nothing (Runner handled it)
+
+### Architecture
+
+The cancellation logic lives in `BaseWorker` to avoid duplication across integrations. Each integration implements only the `kill_infrastructure()` method for their specific infrastructure type.
+
+```
+BaseWorker (src/prefect/workers/base.py)
+├── _setup_cancellation_handling()     # Initialize detection
+├── _poll_for_cancelling_flow_runs()   # Startup poll
+├── _consume_cancellation_events()     # WebSocket consumer
+├── _schedule_force_cancellation()     # Schedule delayed check
+├── _force_cancel_flow_run()           # Terminate and mark cancelled
+└── kill_infrastructure()              # Abstract method
+
+KubernetesWorker
+└── kill_infrastructure()              # Delete K8s Job
+
+DockerWorker
+└── kill_infrastructure()              # Stop Docker container
+
+ECSWorker
+└── kill_infrastructure()              # Stop ECS task
+```
+
+## Rollout Strategy
+
+| Phase | `PREFECT_WORKER_ENABLE_CANCELLATION` | Behavior |
+|-------|--------------------------------------|----------|
+| **Initial release** | `false` (default) | Feature disabled, current behavior |
+| **Opt-in testing** | `true` (user sets) | Feature enabled for testing |
+| **Future release** | `true` (new default) | Feature enabled by default |
+
+Users opt-in by setting the environment variable, allowing them to test the feature before it becomes the default.
+
+## Trade-offs
+
+### On-Cancellation Hooks
+
+When the Worker force-cancels a flow run, `on_cancellation` hooks will not execute because:
+- The Worker doesn't have access to the flow code
+- Hooks run in the flow's execution context with its dependencies
+- Force-cancellation terminates the process immediately
+
+Users who require hooks to always run can:
+- Set a longer grace period to give the Runner more time
+- Set grace period to `-1` to disable worker-side cancellation entirely (accepting that stuck flows may remain in `CANCELLING` indefinitely)
+
+### Duplicate Cancellation Attempts
```

Comment:
Is there any consideration needed for scenarios where more than one worker is running for a work pool?